dissect/target/plugin.py

```diff
@@ -71,7 +63,111 @@ class OperatingSystem(StrEnum):
     WINDOWS = "windows"
 
 
-def export(*args, **kwargs) -> Callable:
+@dataclass(frozen=True, eq=True)
+class PluginDescriptor:
+    module: str
+    qualname: str
+    namespace: str
+    path: str
+    findable: bool
+    functions: list[str]
+    exports: list[str]
+
+
+@dataclass(frozen=True, eq=True)
+class FunctionDescriptor:
+    name: str
+    namespace: str
+    path: str
+    exported: bool
+    internal: bool
+    findable: bool
+    output: str | None
+    method_name: str
+    module: str
+    qualname: str
+
+
+@dataclass(frozen=True, eq=True)
+class FailureDescriptor:
+    module: str
+    stacktrace: list[str]
+
+
+PluginDescriptorLookup = TypedDict(
+    "PluginDescriptorLookup",
+    {
+        None: dict[str, PluginDescriptor],
+        "__os__": dict[str, PluginDescriptor],
+        "__child__": dict[str, PluginDescriptor],
+    },
+)
+
+
+FunctionDescriptorLookup = TypedDict(
+    "FunctionDescriptorLookup",
+    {
+        None: dict[str, dict[str, FunctionDescriptor]],
+        "__os__": dict[str, dict[str, FunctionDescriptor]],
+        "__child__": dict[str, dict[str, FunctionDescriptor]],
+    },
+)
+
+
+PluginRegistry = TypedDict(
+    "PluginRegistry",
+    {
+        "__plugins__": PluginDescriptorLookup,
+        "__functions__": FunctionDescriptorLookup,
+        "__ostree__": dict[str, PluginDescriptor],
+        "__failed__": list[FailureDescriptor],
+    },
+)
+
+PLUGINS: PluginRegistry = {
+    # Plugin descriptor lookup
+    # {"<plugin index>": {"<module_path>": PluginDescriptor}}
+    "__plugins__": {
+        # All regular plugins
+        # {"<module_path>": PluginDescriptor}
+        None: {},
+        # All OS plugins
+        # {"<module_path>": PluginDescriptor}
+        "__os__": {},
+        # All child plugins
+        # {"<module_path>": PluginDescriptor}
+        "__child__": {},
+    },
+    # Function descriptor lookup
+    # {"<plugin index>": {"<function_name>": {"<module_path>": FunctionDescriptor}}}
+    "__functions__": {
+        # All regular plugins
+        # {"<function_name>": {"<module_path>": FunctionDescriptor}}
+        None: {},
+        # All OS plugins
+        # {"<function_name>": {"<module_path>": FunctionDescriptor}}
+        "__os__": {},
+        # All child plugins
+        # {"<function_name>": {"<module_path>": FunctionDescriptor}}
+        "__child__": {},
+    },
+    # OS plugin tree
+    # {"<module_part>": {"<module_part>": PluginDescriptor}}
+    "__ostree__": {},
```

Comment:
This also feels like derived state (i.e. more like a cache) so perhaps it is better to memoize the tree

Reply:
This is indeed a cache, but the important difference between this and memoization is that memoization would need to occur on every first execution, whereas with this we can benefit from a pre-cached index if we deserialise the plugin index.